test/runtests.jl

```diff
@@ -1,6 +1,169 @@
 using L2OALM
 using Test
+using Lux
+using Optimisers
+using MLUtils
+using KernelAbstractions
+using ExaModels
+using BatchNLPKernels
+using CUDA
+
+using PowerModels
+PowerModels.silence()
+using PGLib
+
+using Random
+using Statistics
+
+import GPUArraysCore: @allowscalar
+
+const BNK = BatchNLPKernels
+
+include("power.jl")
 
 @testset "L2OALM.jl" begin
-    # Write your tests here.
+    function feed_forward_builder(
+        num_p::Integer,
+        num_y::Integer,
+        hidden_layers::AbstractVector{<:Integer};
+        activation = relu,
+    )
+        """
+        Builds a Chain of Dense layers with Lux
+        """
+        # Combine all layers: input size, hidden sizes, output size
+        layer_sizes = [num_p; hidden_layers; num_y]
+
+        # Build up a list of Dense layers
+        dense_layers = Any[]
+        for i = 1:(length(layer_sizes)-1)
+            if i < length(layer_sizes) - 1
+                # Hidden layers with activation
+                push!(dense_layers, Dense(layer_sizes[i], layer_sizes[i+1], activation))
+            else
+                # Final layer with no activation
+                push!(dense_layers, Dense(layer_sizes[i], layer_sizes[i+1]))
+            end
+        end
+
+        return Chain(dense_layers...)
+    end
+
+    function test_alm_training(;
+        filename = "pglib_opf_case14_ieee.m",
+        dev_gpu = gpu_device(),
+        backend = CPU(),
+        batch_size = 32,
+        dataset_size = 3200,
+        rng = Random.default_rng(),
+        T = Float64,
+    )
+        model, nbus, ngen, blines =
+            create_parametric_ac_power_model(filename; backend = backend, T = T)
+        bm_train = BNK.BatchModel(model, batch_size, config = BNK.BatchModelConfig(:full))
+        bm_test = BNK.BatchModel(model, dataset_size, config = BNK.BatchModelConfig(:full))
```

Comment:
```suggestion
        bm_train = BNK.BatchModel(model, batch_size, config = BNK.BatchModelConfig(:viol_grad))
        bm_test = BNK.BatchModel(model, dataset_size, config = BNK.BatchModelConfig(:viol_grad))
```

`viol_grad` suffices, just avoids the hessian storage